File: 012_ThreeSum.cpp
// Topic: 3Sum
// Description: Return all unique triplets that sum to zero without duplicates.
// Approach: Sort and use two-pointer for each fixed i, skip duplicates.

#include <bits/stdc++.h>
using namespace std;

class Solution
{
   public:
    vector<vector<int>> threeSum(vector<int>& nums)
    {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        for (int i = 0; i < (int)nums.size(); ++i)
        {
            if (i && nums[i] == nums[i - 1])
                continue;
            int target = -nums[i];
            int l = i + 1, r = (int)nums.size() - 1;
            while (l < r)
            {
                int s = nums[l] + nums[r];
                if (s == target)
                {
                    res.push_back({nums[i], nums[l], nums[r]});
                    while (l < r && nums[l] == nums[l + 1]) ++l;
                    while (l < r && nums[r] == nums[r - 1]) --r;
                    ++l;
                    --r;
                }
                else if (s < target)
                    ++l;
                else
                    --r;
            }
        }
        return res;
    }
};

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    vector<vector<int>> tests = {{-1, 0, 1, 2, -1, -4}, {0, 0, 0, 0}};
    for (size_t t = 0; t < tests.size(); ++t)
    {
        cout << "Test " << (t + 1) << ": ";
        auto a = tests[t];
        for (size_t i = 0; i < a.size(); ++i)
        {
            if (i)
                cout << ",";
            cout << a[i];
        }
        cout << " -> \n";
        auto res = Solution().threeSum(a);
        for (auto& trip : res)
        {
            for (int i = 0; i < 3; ++i)
            {
                if (i)
                    cout << ",";
                cout << trip[i];
            }
            cout << "\n";
        }
        cout << "\n";
    }
}

/*
Time Complexity: O(n^2)
Space Complexity: O(log n) for sort or O(n) depending on sort implementation
*/

File: 012_ThreeSum.py
# Topic: 3Sum
# Description: Return all unique triplets that sum to zero without duplicates.
# Approach: Sort and use two-pointer for each fixed i, skip duplicates.

from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            target = -nums[i]
            l, r = i + 1, len(nums) - 1
            while l < r:
                s = nums[l] + nums[r]
                if s == target:
                    res.append([nums[i], nums[l], nums[r]])
                    while l < r and nums[l] == nums[l + 1]:
                        l += 1
                    while l < r and nums[r] == nums[r - 1]:
                        r -= 1
                    l += 1
                    r -= 1
                elif s < target:
                    l += 1
                else:
                    r -= 1
        return res


if __name__ == "__main__":
    tests = [[-1, 0, 1, 2, -1, -4], [0, 0, 0, 0]]
    sol = Solution()
    for i, nums in enumerate(tests):
        print(f"Test {i + 1}: {nums} ->")
        result = sol.threeSum(nums)
        for trip in result:
            print(','.join(map(str, trip)))
        print()

"""
Time Complexity: O(n^2)
Space Complexity: O(log n) for sort or O(n) depending on sort implementation
"""
